### ISSUE TITLE
osx installation

### ISSUE THREAD

**Issue by jpnayak** [sanguinariojoe/aquagpusph#11]
Any suggestion how to compile and install on osx? Have installed all of the dependencies with help of homebrew.

**Comment by sanguinariojoe** [comment]
Theoretically, you can compile AQUAgpusph in OSX in the same way than in Linux. Check https://gitlab.com/sanguinariojoe/aquagpusph/wikis/install#compiling

**Comment by jpnayak** [comment]
But a Reliable Guide is Essential, particularly when one have to install dependencies through homebrew. You know apt-get has no luck in osx sierra.
I've tried but failed to compile.
here is the error log of the compilation.
[CMakeOutput.log](https://github.com/sanguinariojoe/aquagpusph/files/1445086/CMakeOutput.log)

**Comment by jpnayak** [comment]
Well anyhow, I've reordered the compiler detection to detect the homebrew directory first, as all of these dependencies are installed via this. Everything is now detected and configures except "/aquagpusph/include/sphPrerequisites.h:47:10:fatal error: 'config.h' file not found" and hence the compilation process stops at once.

Here's the screenshot of my terminal showing the error:
<img width="1280" alt="screen shot 2017-12-08 at 2 56 12 pm" src="https://user-images.githubusercontent.com/15700234/33760098-9dbf2100-dc2a-11e7-9521-17a452814187.png">

And here's the cmake output log:
[CMakeOutput.log](https://github.com/sanguinariojoe/aquagpusph/files/1542179/CMakeOutput.log)

**Comment by sanguinariojoe** [comment]
The best solution to fix that is using GCC:

```
cmake -DCMAKE_C_COMPILER=/usr/bin/gcc -DCMAKE_CXX_COMPILER=/usr/bin/g++ ...
```

Alternatively, you may try to patch that replacing in CMakeLists.txt the following:

```
# ===================================================== #
# config.h stuff                                        #
# ===================================================== #
if(CMAKE_COMPILER_IS_GNUCXX)
    include(cMake/ConfigureChecks.cmake)
    configure_file(config.h.cmake ${CMAKE_CURRENT_BINARY_DIR}/config.h)
    add_definitions(-DHAVE_CONFIG_H)
    add_definitions(-Wno-write-strings)
    add_definitions(-Wno-deprecated)
    INCLUDE_DIRECTORIES(${CMAKE_CURRENT_BINARY_DIR})
    # get linker errors as soon as possible and not at runtime e.g. for modules
    if(UNIX)
    #    SET(CMAKE_SHARED_LINKER_FLAGS "-Wl,--no-undefined")
    endif(UNIX)
endif(CMAKE_COMPILER_IS_GNUCXX)
```
by:

```
# if(CMAKE_COMPILER_IS_GNUCXX)
    include(cMake/ConfigureChecks.cmake)
    configure_file(config.h.cmake ${CMAKE_CURRENT_BINARY_DIR}/config.h)
    add_definitions(-DHAVE_CONFIG_H)
    add_definitions(-Wno-write-strings)
    add_definitions(-Wno-deprecated)
    INCLUDE_DIRECTORIES(${CMAKE_CURRENT_BINARY_DIR})
    # get linker errors as soon as possible and not at runtime e.g. for modules
    if(UNIX)
    #    SET(CMAKE_SHARED_LINKER_FLAGS "-Wl,--no-undefined")
    endif(UNIX)
# endif(CMAKE_COMPILER_IS_GNUCXX)
```

That's such a super-bad idea, because the flags shall be customized for each compiler...

**Comment by jpnayak** [comment]
Yeah, I've replaced the config.h stuff in cmakelists.txt and compiled it with my brew gcc and g++ paths. But another problem has been preventing the compilation.
It says that `fatal error: CL/cl.h: No such file or directory
compilation terminated.`
Seems it couldn't find out where opencl lives. I've tried to search it manually by `find / -iname cl.h 2>/dev/null` and it returns every possible directory where opencl is found and it seems I've already `cl.h` on my machine.

[CMakeError.log](https://github.com/sanguinariojoe/aquagpusph/files/1544515/CMakeError.log) 
[CMakeOutput.log](https://github.com/sanguinariojoe/aquagpusph/files/1544517/CMakeOutput.log)
<img width="1280" alt="screen shot 2017-12-09 at 2 40 12 pm" src="https://user-images.githubusercontent.com/15700234/33794300-073db582-dcef-11e7-94f6-fa43d2c4fb66.png">

**Comment by sanguinariojoe** [comment]
Some Apple idiot decided to install cl.h in a folder called OpenCL instead of CL. The easiest way to fix that problem is creating a `include/CL` subfolder (inside AQUAgpusph root path), and paste inside your `cl.h` file.

**Comment by jpnayak** [comment]
Well, I've copied the `cl.h` file and tried to compile. every configure and build seems to run good but linking fails.
It says `Undefined symbols for architecture x86_64` and generates error.
Here are the log files:
[terminal bash.txt](https://github.com/sanguinariojoe/aquagpusph/files/1545298/terminal.bash.txt)
[CMakeError.log](https://github.com/sanguinariojoe/aquagpusph/files/1545299/CMakeError.log)
[CMakeOutput.log](https://github.com/sanguinariojoe/aquagpusph/files/1545300/CMakeOutput.log)

**Comment by jpnayak** [comment]
I hope you have found some solution to the issue by now.

**Comment by sanguinariojoe** [comment]
Ops! I forgot about you! Sorry!

It seems that your compiler is trying to statically link all the stuff. However both server and client libraries have shared symbols (the idea is removing such cross-references in the future).

You can try to merge `src/CMakeLists.txt` and `src/CalcServer/CMakeLists.txt` to produce a monolithic AQUAgpusph.so library, instead of AQUAgpusphClient.so and AQUAgpusphServer.so
